ipfix/decoder.go

```diff
@@ -96,75 +98,115 @@ func NewDecoder(raddr net.IP, b []byte) *Decoder {
 
 // Decode decodes the IPFIX raw data
 func (d *Decoder) Decode(mem MemCache) (*Message, error) {
-	var (
-		msg = new(Message)
-		err error
-	)
+	var msg = new(Message)
 
 	// IPFIX Message Header decoding
-	if err = msg.Header.unmarshal(d.reader); err != nil {
+	if err := msg.Header.unmarshal(d.reader); err != nil {
 		return nil, err
 	}
 	// IPFIX Message Header validation
-	if err = msg.Header.validate(); err != nil {
+	if err := msg.Header.validate(); err != nil {
 		return nil, err
 	}
 
 	// Add source IP address as Agent ID
 	msg.AgentID = d.raddr.String()
 
+	// In case there are multiple non-fatal errors, collect them and report all of them.
+	// The rest of the received sets will still be interpreted, until a fatal error is encountered.
+	// A non-fatal error is for example an illegal data record or unknown template id.
+	var decodeErrors []error
 	for d.reader.Len() > 4 {
+		if err := d.decodeSet(mem, msg); err != nil {
+			switch err.(type) {
+			case nonfatalError:
+				decodeErrors = append(decodeErrors, err)
+			default:
+				return nil, err
+			}
+		}
+	}
 
-		setHeader := new(SetHeader)
-		setHeader.unmarshal(d.reader)
+	return msg, combineErrors(decodeErrors...)
+}
 
-		if setHeader.Length < 4 {
-			return nil, io.ErrUnexpectedEOF
-		}
+type nonfatalError error
 
-		switch {
-		case setHeader.SetID == 2:
-			// Template set
-			tr := TemplateRecord{}
-			tr.unmarshal(d.reader)
-			mem.insert(tr.TemplateID, d.raddr, tr)
-		case setHeader.SetID == 3:
-			// Option set
-			tr := TemplateRecord{}
-			tr.unmarshalOpts(d.reader)
-			mem.insert(tr.TemplateID, d.raddr, tr)
-		case setHeader.SetID >= 4 && setHeader.SetID <= 255:
-			// Reserved
-		default:
-			// data
-			tr, ok := mem.retrieve(setHeader.SetID, d.raddr)
-			if !ok {
-				select {
-				case rpcChan <- RPCRequest{
-					ID: setHeader.SetID,
-					IP: d.raddr,
-				}:
-				default:
-				}
-				return msg, fmt.Errorf("%s unknown ipfix template id# %d",
-					d.raddr.String(),
-					setHeader.SetID,
-				)
+func (d *Decoder) decodeSet(mem MemCache, msg *Message) error {
+	startCount := d.reader.ReadCount()
+
+	setHeader := new(SetHeader)
+	if err := setHeader.unmarshal(d.reader); err != nil {
+		return err
+	}
+	if setHeader.Length < 4 {
+		return io.ErrUnexpectedEOF
+	}
+
+	var tr TemplateRecord
+	var err error
+	// This check is somewhat redundant with the switch-clause below, but the retrieve() operation should not be executed inside the loop.
+	if setHeader.SetID > 255 {
+		var ok bool
+		tr, ok = mem.retrieve(setHeader.SetID, d.raddr)
+		if !ok {
+			select {
+			case rpcChan <- RPCRequest{
+				ID: setHeader.SetID,
+				IP: d.raddr,
+			}:
+			default:
 			}
+			err = nonfatalError(fmt.Errorf("%s unknown ipfix template id# %d. Known ids: %v",
+				d.raddr.String(),
+				setHeader.SetID,
+				mem.allSetIds(),
```

Comment:
As you mentioned allSetIds is good for debugging purpose not production. it's fine to keep the allSetIds() code in case someone needs to debug but not here.
Let's say if a network has more than 100 routers if an unknown ipfix case happens then we will get many SetID at log file for different routers.